File: 최단거리/BellmanFord/BellmanFord.js
class Edge {
    constructor(from, to, cost){
        this.from = from;
        this.to = to;
        this.cost = cost;
    }
}

let dist = [];
let edgeList = [];

function bellmanFord(v, start){
    dist[start] = 0;

    for(let i=0; i<v-1; i++){
        for(let e of edgeList){
            if(dist[e.from] === Number.MAX_SAFE_INTEGER) continue;

            if(dist[e.to] > dist[e.from] + e.cost)
                dist[e.to] = dist[e.from] + e.cost;
        }
    }

    for(let e of edgeList){
        if(dist[e.from] === Number.MAX_SAFE_INTEGER) continue;

        if(dist[e.to] > dist[e.from] + e.cost)
            return true;
    }

    return false;
}

function main() {
    const fs = require('fs');
    const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

    const [v, e] = input[0].split(' ').map(Number);

    dist = new Array(v+1).fill(Number.MAX_SAFE_INTEGER);
    edgeList = new Array(e).fill().map(() => new Edge(0,0,0));

    for(let i=0; i<e; i++){
        const [from, to, cost] = input[i+1].split(' ').map(Number);
        edgeList[i] = new Edge(from, to, cost);
    }

    const start = Number(input[e+1]);
    let haveNegativeCycle = bellmanFord(v, start);
}

File: 최단거리/BellmanFord/BellmanFord.cpp
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

struct edge {
  int from, to, cost;
};

int v, e;
vector<long long int> dist;
vector<edge> edgeList;

bool bellmanFord(int start) {
  dist[start] = 0;

  // v-1번 탐색
  for (int i = 0; i < v - 1; i++) {
    for (edge e : edgeList) {
      // 간선의 시작점이 탐색불가
      if (dist[e.from] == LLONG_MAX)
        continue;

      if (dist[e.to] > dist[e.from] + e.cost) {
        dist[e.to] = dist[e.from] + e.cost;
      }
    }
  }

  // 음의 사이클 검사
  for (edge e : edgeList) {
    if (dist[e.from] == LLONG_MAX)
      continue;

    if (dist[e.to] > dist[e.from] + e.cost) {
      return true;
    }
  }

  return false;
}

int main() {
  cin >> v >> e;

  dist.resize(v + 1, LLONG_MAX);
  edgeList.resize(e);

  for (int i = 0; i < e; i++) {
    cin >> edgeList[i].from >> edgeList[i].to >> edgeList[i].cost;
  }

  int start;
  cin >> start;

  bool haveNegativeCycle = bellmanFord(start);

  return 0;
}
